libqtile/core/manager.py

```diff
@@ -173,6 +175,14 @@ def load_config(self, initial=False) -> None:
         if self.config.reconfigure_screens:
             hook.subscribe.screen_change(self.cmd_reconfigure_screens)
 
+        # To ensure compliance with X11's window layering spec
+        # We need to restack fullscreen windows if they lose focus
+        hook.subscribe.client_focus(self.check_stacking)
```

Comment:
focus_change might be more appropriate, as it also fires when focus moves to nothing?

If this is X11-specific, could this be moved into the x11 core? It may not need to be hooked - there might be a good place to put it near the code that actually does the focussing.

I'm also a bit concerned that this may have unexpected side effects (and become difficult to debug). We don't have any core core code (only configurables code) that depends on hooks. Note that cmd_reload_config clears hooks, so currently hooks are a more user-oriented functionality. This hook here would get unsubscribed upon config reload and wouldn't get rehooked.

Reply:
It's X11 specific in that there are rules/guides on priority for stacking windows. Any reason why Wayland wouldn't apply similar guides? i.e. presumably you'll still want the ability to keep windows above others and have fullscreen windows on top?

Reply:
You would still want those things but clients are not aware of their relative positions, so there is no informing them of their position, and therefore not really anything to do upon focus change. It's also possible I've misunderstood a bit about what this part is doing because I wasn't so sure what `self.check_stacking` was for. Because the wayland backend is what determines Z axis at every level (by choosing the order to render windows in, and where input events get sent), it doesn't need a hook to respond to its own actions. So it's more the implementation detail that is x11 specific, rather than the concept

Reply:
Happy to leave this is as. If there is an issue with the implementation it will come up pretty soon as a bug, and we can fix it then